scripts/nn/layers/softmax2d.dml

```diff
@@ -0,0 +1,69 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+/*
+ * 2D - Softmax classifier layer used for image segmentation.
+ */
+
+forward = function(matrix[double] scores, int C, int H, int W)
+    return (matrix[double] probs) {
+  /*
+   * Computes the forward pass for a softmax classifier.  The inputs
+   * are interpreted as unnormalized, log-probabilities for each of
+   * N examples, and the softmax function transforms them to normalized
+   * probabilities.
+   *
+   * This can be interpreted as a generalization of the sigmoid
+   * function to multiple classes.
+   *
+   *   `probs_ij = e^scores_ij / sum(e^scores_i)`
+   *
+   * Inputs:
+   *  - scores: Inputs, of shape (N, C*HW).
+   *  - C: Number of channels or output labels.
+   *  - H: Image height
+   *  - W: Image width
+   *
+   * Outputs:
+   *  - probs: Outputs, of shape (N, C*HW).
+   */
+  # For numerical stability, we subtract the max score of an example from all scores for that
+  # example.  This is equivalent to the original formulation:
+  # e^scores_i / sum(e^scores_i) == C*e^scores_i / C*sum(e^scores_i)
+  #                              == e^(scores_i+log(C)) / sum(e^(scores_i+log(C))
+  # set log(C) = -max(scores_i):
+  #                              == e^(scores_i-max(scores_i)) / sum(e^(scores_i-max(scores_i))
+  N = nrow(scores)
+  scores = matrix(scores, rows=N*C, cols=H*W)
```

Comment:
Excellent, yes, that is the correct axis permutation that I am looking for.  The actual builtin `transpose4d` will be the better way to go in general, it just depends on how long that will take as compared to the DML utilities, and how soon we want the community to be able to work on image segmentation tasks.  Even better than all of that, generalizing our builtin `matrix` type to a `tensor` type with the same functionality as NumPy arrays would be the best option.

@mboehm7 Realistically, what would the path to adding a `tensor` type into the engine look like?